### PR TITLE
Allow TernaryLogic to optimize down to BlendVariableMask where appropriate

### PR DESCRIPTION
**Pull request by tannergooding**
Given the following code:
```csharp
public static Vector512<float> M(Vector512<float> x, Vector512<float> y)
{
    return Vector512.ConditionalSelect(Vector512.Equals(x, y), x, y);
}
```

###Before:

```asm
; Method Program:M(System.Runtime.Intrinsics.Vector512`1[float],System.Runtime.Intrinsics.Vector512`1[float]):System.Runtime.Intrinsics.Vector512`1[float] (FullOpts)
G_M44221_IG01:  ;; offset=0x0000
       vzeroupper 
						;; size=3 bbWeight=1 PerfScore 1.00

G_M44221_IG02:  ;; offset=0x0003
       vmovups  zmm0, zmmword ptr [rdx]
       vmovups  zmm1, zmmword ptr [r8]
       vcmpps   k1, zmm0, zmm1, 0
       vpmovm2d zmm2, k1
       vpternlogd zmm2, zmm0, zmm1, -54
       vmovups  zmmword ptr [rcx], zmm2
       mov      rax, rcx
						;; size=41 bbWeight=1 PerfScore 14.75

G_M44221_IG03:  ;; offset=0x002C
       vzeroupper 
       ret      
						;; size=4 bbWeight=1 PerfScore 2.00
; Total bytes of code: 48
```

### After:

```asm
; Method Program:M(System.Runtime.Intrinsics.Vector512`1[float],System.Runtime.Intrinsics.Vector512`1[float]):System.Runtime.Intrinsics.Vector512`1[float] (FullOpts)
G_M44221_IG01:  ;; offset=0x0000
       vzeroupper 
						;; size=3 bbWeight=1 PerfScore 1.00

G_M44221_IG02:  ;; offset=0x0003
       vmovups  zmm0, zmmword ptr [rdx]
       vmovups  zmm1, zmmword ptr [r8]
       vcmpps   k1, zmm0, zmm1, 0
       vblendmps zmm0 k1, zmm1, zmm0
       vmovups  zmmword ptr [rcx], zmm0
       mov      rax, rcx
						;; size=35 bbWeight=1 PerfScore 13.75

G_M44221_IG03:  ;; offset=0x0026
       vzeroupper 
       ret      
						;; size=4 bbWeight=1 PerfScore 2.00
; Total bytes of code: 42
```

### CODE REVIEW

**Comment by stephentoub** [comment]
You'd mentioned to me such lowering is already being done in some cases. For my edification, how does the scenerios this newly covers differ from what was already covered? Is this about 512 specifically now?

**Comment by tannergooding** [comment]
This is specifically about handling `TYP_MASK` inputs that exist in AVX512 or later, since that hardware has a special mask register (`k1-k7`) that is produced by comparisons and can be used to do masking based operations such as this.

Basically ensuring that the codegen is optimal for V512, where it was already handling this for V128/V256.

**Comment by tannergooding** [comment]
In the diffs just linked, you can see an example of this where we can remove a few `vpmovm2b` and `vpternlogd` instructions in favor of consuming `k1` directly with `vpblendmb`. For V128/V256 the `vpcmpub` instruction produces an `xmm/ymm` register and so `vpternlogd` is the better choice since it can be consumed as is.
```diff
-       vpmovm2b zmm2, k1
-       vxorps   ymm3, ymm3, ymm3
-       vpcmpub  k1, zmm0, zmm3, 1
-       vpmovm2b zmm3, k1
-       vpternlogd zmm3, zmm0, zmm1, -54
-       vpcmpub  k1, zmm0, zmm1, 1
-       vpmovm2b zmm4, k1
-       vpternlogd zmm4, zmm0, zmm1, -54
-       vpternlogd zmm2, zmm3, zmm4, -54
-       vmovups  zmmword ptr [rdi], zmm2
+       vxorps   ymm2, ymm2, ymm2
+       vpcmpub  k2, zmm0, zmm2, 1
+       vpblendmb zmm2 {k2}, zmm1, zmm0
+       vpcmpub  k2, zmm0, zmm1, 1
+       vpblendmb zmm0 {k2}, zmm1, zmm0
+       vpblendmb zmm0 {k1}, zmm0, zmm2
+       vmovups  zmmword ptr [rdi], zmm0
```

**Review by kunalspathak — APPROVED** [review]
LGTM

File: src/coreclr/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -2689,6 +2694,144 @@ GenTree* Lowering::LowerHWIntrinsicCndSel(GenTreeHWIntrinsic* node)
...
+
+                if (!condition->OperIsHWIntrinsic(NI_AVX512F_ConvertMaskToVector))
+                {
+                    break;
+                }
+
+                node->ResetHWIntrinsicId(NI_AVX512F_BlendVariableMask, comp, selectFalse, selectTrue,
+                                         condition->AsHWIntrinsic()->Op(1));
+
+                BlockRange().Remove(op1);
```
Comment (EgorBo) [inline src/coreclr/jit/lowerxarch.cpp:2819]: is it possible that you're deleting nodes with side-effects?

**Review by EgorBo — APPROVED** [review]

File: src/coreclr/jit/lowerxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/lowerxarch.cpp:2819]: It's explicitly checked to be the hwintrinsic node ConvertVectorToMask, so it shouldn't ever had any side effects

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -16,6 +16,7 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 #endif
 
 #include "allocacheck.h" // for alloca
+#include "hwintrinsic.h"
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/coreclr/jit/morph.cpp:19]: Ah, no. I had started by doing the change in morph and realized it wouldn't be a good fit due to the parameter swapping and needing to deal with potential side effects from commas/etc

File: src/coreclr/jit/lowerxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/lowerxarch.cpp:2819]: Ah, you know what, though. This is deleting the wrong thing and should be removing `condition` not `op1`. Let me fix that

**Pull request merged**